enhancements/machine-api/opt-out-machine-health-checking.md

```diff
@@ -0,0 +1,378 @@
+---
+title: opt-out-machine-health-checking
+authors:
+  - "@JoelSpeed"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-05-18
+last-updated: 2020-05-18
+status: implementable
+see-also:
+  - "/enhancements/machine-api/machine-health-checking.md"  
+---
+
+# Opt-Out Machine Health Checking
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This proposal explores the option to install Machine Health Checking by default within new OpenShift clusters.
+This would allow clusters to automatically heal themselves when machines become unhealthy.
+
+## Motivation
+
+Improve the user experience of OpenShift by automatically healing Machines without requiring extra configuration from users.
+
+### Goals
+
+- Provide a sensible default MachineHealthCheck on new OpenShift clusters
+- Allow worker Machines to recover automatically
+- Allow users to opt-out of health checking
+
+### Non-Goals
+
+- Force MachineHealthChecks to always be present on clusters
+- Provide MachineHealthChecks as part of upgrade paths
+
+## Proposal
+
+On cluster creation, the installer should create a MachineHealthCheck that targets all MachineSets created by the installer.
+It will not target Control-Plane Machines created by the installer.
+
+This HealthCheck will mark machines unhealthy based on the `Ready` condition on the Machine's Node.
+
+It will also specify a sensible default for the `maxUnhealthy` field which will prevent remediation of unhealthy
+Machines once a certain percentage of the cluster is considered unhealthy.
+
+Should cluster administrators not be happy with the MachineHealthCheck, they will be free to modify or delete it as they wish.
+
+```yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineHealthCheck
+metadata:
+  name: machine-unready-5m
+  namespace: openshift-machine-api
+spec:
+  nodeStartupTimeout: 10m # This is the default value
+  maxUnhealthy: 40%
+  selector:
+    matchLabels:
+      machine.openshift.io/cluster-api-machine-role: worker
+      machine.openshift.io/cluster-api-machine-type: worker
+    matchExpressions:
+    - key: machine.openshift.io/cluster-api-machineset
+      operator: In
+      values:
+      - <machineset-name-a>
+      - <machineset-name-b>
+      - <machineset-name-c>
+  unhealthyConditions:	 
+  - type:    Ready
+    timeout: 300s
+    status: "False"
+  - type:    Ready
+    timeout: 300s
+    status: Unknown
+```
+
+### Implementation Details
+
+A new MachineHealthCheck manifest will be added to the installer resources.
+This MachineHealthCheck should **not** be managed by the Cluster Version Operator,
+this would prevent users from adjusting it to suit their particular needs.
```

Comment:
i get that the CVO should not manage the details of MHC deployment, but would it manage the version information about the MHC?

Reply:
Do you mean the `apiVersion` of the `MachineHealthCheck`? This would be upgraded by the API server as the cluster gets upgraded so it doesn't need managing per se. The only reason we would need management were be if we changed the API completely and moved to a new API group, in which case we would have to deprecate the old one

Reply:
i was thinking more about image and binary version stuff, but perhaps i am ascribing behavior to the CVO that does not exist.

Reply:
This is just about deploying a single MachineHealthCheck CR, the CVO will still manage the MHC contorller indirectly through the MAO